File: fastapi-backend/app/services/ollama_inference_service.py
import httpx
import time
import os
from typing import Dict, Literal, Optional

class OllamaInferenceService:
    """
    Async Ollama inference service for FastAPI backend.
    Uses httpx for non-blocking HTTP calls.
    """
    
    def __init__(self):
        self.base_url = os.getenv("OLLAMA_BASE_URL", "http://localhost:11434")
        self.model = os.getenv("OLLAMA_MODEL", "phi3:mini")
        self.generate_url = f"{self.base_url}/api/generate"
        self.tags_url = f"{self.base_url}/api/tags"
        
        # Standard inference parameters
        self.inference_params = {
            "temperature": 0.1,
            "num_predict": 150,
            "top_k": 10,
            "top_p": 0.5
        }
        
        print(f"✓ Ollama Inference Service initialized: {self.model} at {self.base_url}")
    
    def _build_phi3_prompt(self, system_prompt: str, user_prompt: str) -> str:
        """
        Build phi3:mini native chat template format.
        
        Args:
            system_prompt: System instructions
            user_prompt: User query
            
        Returns:
            Formatted prompt with phi3 tags
        """
        return f"""<|system|>
{system_prompt}
<|end|>

<|user|>
{user_prompt}
<|end|>

<|assistant|>"""
    
    async def health_check(self) -> Dict:
        """
        Verify Ollama is reachable and model is loaded.
        
        Returns:
            Dict with healthy (bool), message (str), model (str)
        """
        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                response = await client.get(self.tags_url)
                
                if response.status_code != 200:
                    return {
                        "healthy": False,
                        "message": f"Ollama returned status {response.status_code}",
                        "model": self.model
                    }
                
                data = response.json()
                
                if "models" not in data:
                    return {
                        "healthy": False,
                        "message": "Ollama responded but no models found",
                        "model": self.model
                    }
                
                # Check if our model exists
                model_exists = any(self.model in m.get("name", "") for m in data["models"])
                
                if not model_exists:
                    available = [m.get("name") for m in data["models"]]
                    return {
                        "healthy": False,
                        "message": f"Model {self.model} not found. Available: {', '.join(available)}",
                        "model": self.model
                    }
                
                return {
                    "healthy": True,
                    "message": f"Ollama ready with {self.model}",
                    "model": self.model
                }
                
        except httpx.ConnectError:
            return {
                "healthy": False,
                "message": f"Ollama unreachable at {self.base_url}. Is Ollama running?",
                "model": self.model
            }
        except Exception as e:
            return {
                "healthy": False,
                "message": f"Health check failed: {str(e)}",
                "model": self.model
            }
    
    async def generate_rag_response(
        self,
        context: str,
        query: str,
        mode: Literal["emergency", "deep"] = "emergency"
    ) -> Dict:
        """
        Generate RAG response based on retrieved context.
        
        Args:
            context: Retrieved medical context from vector search
            query: User query
            mode: "emergency" (fast, <150 tokens) or "deep" (detailed, <150 tokens)
            
        Returns:
            Dict with answer, llm_latency_ms, and optional error
        """
        start_time = time.time()
        
        try:
            # Mode-specific system prompts
            if mode == "emergency":
                system_prompt = """You are an emergency clinical decision assistant.
Respond in under 120 words.
Provide only critical actionable recommendation.
Avoid hallucination.
Base reasoning strictly on provided medical context."""
            else:  # deep mode
                system_prompt = """You are an intelligent document analysis assistant.
Answer the user's question directly and concisely using only information from the document.
Do NOT provide a full clinical assessment or summarize the document unless explicitly asked.
Always cite specific parts of the document when referencing information."""
            
            user_prompt = f"""Medical Context:
{context}

Query: {query}

Please answer the query directly using only the provided context."""
            
            prompt = self._build_phi3_prompt(system_prompt, user_prompt)
            
            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.post(
                    self.generate_url,
                    json={
                        "model": self.model,
                        "prompt": prompt,
                        "stream": False,
                        "options": self.inference_params
                    }
                )
                
                llm_latency_ms = round((time.time() - start_time) * 1000, 2)
                
                if response.status_code != 200:
                    return {
                        "answer": "Unable to generate response. Ollama returned an error.",
                        "llm_latency_ms": llm_latency_ms,
                        "error": f"HTTP {response.status_code}"
                    }
                
                data = response.json()
                
                if "response" not in data:
                    return {
                        "answer": "Unable to generate response. Ollama returned malformed data.",
                        "llm_latency_ms": llm_latency_ms,
                        "error": "MALFORMED_RESPONSE"
                    }
                
                return {
                    "answer": data["response"].strip(),
                    "llm_latency_ms": llm_latency_ms
                }
                
        except httpx.ConnectError:
            llm_latency_ms = round((time.time() - start_time) * 1000, 2)
            return {
                "answer": "AI service is currently unavailable. Ollama is not reachable. Please ensure Ollama is running.",
                "llm_latency_ms": llm_latency_ms,
                "error": "OLLAMA_UNREACHABLE"
            }
        except httpx.TimeoutException:
            llm_latency_ms = round((time.time() - start_time) * 1000, 2)
            return {
                "answer": "Request timed out. The AI model took too long to respond.",
                "llm_latency_ms": llm_latency_ms,
                "error": "TIMEOUT"
            }
        except Exception as e:
            llm_latency_ms = round((time.time() - start_time) * 1000, 2)
            return {
                "answer": f"An unexpected error occurred: {str(e)}",
                "llm_latency_ms": llm_latency_ms,
                "error": "INFERENCE_FAILED"
            }
    
    async def generate_naive_response(
        self,
        full_context: str,
        query: str
    ) -> Dict:
        """
        Naive approach: Send full patient history without retrieval optimization.
        Used for comparison benchmarking.
        
        Args:
            full_context: Complete unfiltered patient history
            query: User query
            
        Returns:
            Dict with answer, llm_latency_ms, and optional error
        """
        start_time = time.time()
        
        try:
            system_prompt = "You are a clinical AI assistant. Analyze the patient history and answer the query."
            
            user_prompt = f"""Full Patient History:
{full_context}

Query: {query}

Provide your clinical assessment."""
            
            prompt = self._build_phi3_prompt(system_prompt, user_prompt)
            
            async with httpx.AsyncClient(timeout=15.0) as client:
                response = await client.post(
                    self.generate_url,
                    json={
                        "model": self.model,
                        "prompt": prompt,
                        "stream": False,
                        "options": self.inference_params
                    }
                )
                
                llm_latency_ms = round((time.time() - start_time) * 1000, 2)
                
                if response.status_code != 200:
                    return {
                        "answer": "Unable to generate response.",
                        "llm_latency_ms": llm_latency_ms,
                        "error": f"HTTP {response.status_code}"
                    }
                
                data = response.json()
                
                if "response" not in data:
                    return {
                        "answer": "Malformed response from Ollama.",
                        "llm_latency_ms": llm_latency_ms,
                        "error": "MALFORMED_RESPONSE"
                    }
                
                return {
                    "answer": data["response"].strip(),
                    "llm_latency_ms": llm_latency_ms
                }
                
        except httpx.ConnectError:
            llm_latency_ms = round((time.time() - start_time) * 1000, 2)
            return {
                "answer": "Ollama service unreachable.",
                "llm_latency_ms": llm_latency_ms,
                "error": "OLLAMA_UNREACHABLE"
            }
        except Exception as e:
            llm_latency_ms = round((time.time() - start_time) * 1000, 2)
            return {
                "answer": f"Error: {str(e)}",
                "llm_latency_ms": llm_latency_ms,
                "error": "INFERENCE_FAILED"
            }

# Global instance (initialized in main.py)
ollama_service: Optional[OllamaInferenceService] = None

File: emergency-triage-assistant/backend/src/services/ollamaInferenceService.js
const axios = require('axios');

// Load from environment variables with fallbacks
const OLLAMA_BASE_URL = process.env.OLLAMA_BASE_URL || 'http://localhost:11434';
const OLLAMA_MODEL = process.env.OLLAMA_MODEL || 'phi3:mini';
const OLLAMA_GENERATE_URL = `${OLLAMA_BASE_URL}/api/generate`;
const OLLAMA_TAGS_URL = `${OLLAMA_BASE_URL}/api/tags`;

// Standard inference parameters for all requests
const INFERENCE_PARAMS = {
  temperature: 0.1,
  num_predict: 150,
  top_k: 10,
  top_p: 0.5
};

/**
 * Build phi3:mini native chat template prompt
 * @param {string} systemPrompt - System instructions
 * @param {string} userPrompt - User query
 * @returns {string} Formatted prompt with phi3 tags
 */
const buildPhi3Prompt = (systemPrompt, userPrompt) => {
  return `<|system|>
${systemPrompt}
<|end|>

<|user|>
${userPrompt}
<|end|>

<|assistant|>`;
};

/**
 * Health check: Verify Ollama is reachable and phi3:mini model is loaded
 * @returns {Promise<{healthy: boolean, message: string, model: string}>}
 */
const healthCheck = async () => {
  try {
    // Check if Ollama server is reachable
    const response = await axios.get(OLLAMA_TAGS_URL, { timeout: 5000 });
    
    if (!response.data || !response.data.models) {
      return {
        healthy: false,
        message: 'Ollama server responded but no models found',
        model: OLLAMA_MODEL
      };
    }

    // Check if phi3:mini is available
    const modelExists = response.data.models.some(m => m.name.includes(OLLAMA_MODEL));
    
    if (!modelExists) {
      return {
        healthy: false,
        message: `Model ${OLLAMA_MODEL} not found. Available models: ${response.data.models.map(m => m.name).join(', ')}`,
        model: OLLAMA_MODEL
      };
    }

    return {
      healthy: true,
      message: `Ollama is ready with ${OLLAMA_MODEL}`,
      model: OLLAMA_MODEL
    };
  } catch (error) {
    return {
      healthy: false,
      message: `Ollama unreachable at ${OLLAMA_BASE_URL}: ${error.message}`,
      model: OLLAMA_MODEL
    };
  }
};

/**
 * Emergency triage query: Analyze patient complaint and return structured severity assessment
 * @param {string} complaint - Patient chief complaint
 * @param {Object} vitals - Optional vitals object {bp, pulse, temp, spo2}
 * @returns {Promise<{severity: string, reason: string, actions: string[], responseTime: number}>}
 */
const emergencyTriageQuery = async (complaint, vitals = null) => {
  const startTime = Date.now();

  try {
    const systemPrompt = `You are an Emergency Medicine AI Assistant.
Analyze the patient complaint and vitals to determine triage severity.
Respond in EXACTLY this format:

SEVERITY: [CRITICAL/HIGH/MEDIUM/LOW]
REASON: [One sentence why]
ACTIONS: [Action 1] | [Action 2] | [Action 3]

Be concise. No extra text.`;

    let userPrompt = `Patient Complaint: ${complaint}`;
    
    if (vitals) {
      userPrompt += `\n\nVitals:`;
      if (vitals.bp) userPrompt += `\n- BP: ${vitals.bp}`;
      if (vitals.pulse) userPrompt += `\n- Pulse: ${vitals.pulse} bpm`;
      if (vitals.temp) userPrompt += `\n- Temperature: ${vitals.temp}°F`;
      if (vitals.spo2) userPrompt += `\n- SpO2: ${vitals.spo2}%`;
    }

    userPrompt += `\n\nProvide triage assessment.`;

    const prompt = buildPhi3Prompt(systemPrompt, userPrompt);

    const response = await axios.post(OLLAMA_GENERATE_URL, {
      model: OLLAMA_MODEL,
      prompt: prompt,
      stream: false,
      options: INFERENCE_PARAMS
    }, {
      timeout: 10000
    });

    const responseTime = Date.now() - startTime;

    if (!response.data || !response.data.response) {
      throw new Error('Ollama returned malformed response');
    }

    const text = response.data.response.trim();

    // Parse structured response
    const severityMatch = text.match(/SEVERITY:\s*(\w+)/i);
    const reasonMatch = text.match(/REASON:\s*(.+?)(?=\n|ACTIONS:|$)/is);
    const actionsMatch = text.match(/ACTIONS:\s*(.+?)$/is);

    const severity = severityMatch ? severityMatch[1].toUpperCase() : 'MEDIUM';
    const reason = reasonMatch ? reasonMatch[1].trim() : 'Unable to determine';
    const actions = actionsMatch 
      ? actionsMatch[1].split('|').map(a => a.trim()).filter(a => a.length > 0)
      : ['Assess patient', 'Monitor vitals', 'Consult physician'];

    return {
      severity,
      reason,
      actions,
      responseTime
    };

  } catch (error) {
    const responseTime = Date.now() - startTime;
    
    if (error.code === 'ECONNREFUSED' || error.message.includes('unreachable')) {
      return {
        severity: 'UNKNOWN',
        reason: 'Ollama service is currently unavailable. Please ensure Ollama is running.',
        actions: ['Manual triage required', 'Check Ollama service status'],
        responseTime,
        error: 'OLLAMA_UNREACHABLE'
      };
    }

    return {
      severity: 'UNKNOWN',
      reason: `AI analysis failed: ${error.message}`,
      actions: ['Manual triage required', 'Review patient complaint directly'],
      responseTime,
      error: 'INFERENCE_FAILED'
    };
  }
};

/**
 * Multi-turn conversation for patient chatbot interface
 * @param {Array<{role: string, content: string}>} conversationHistory - Array of messages
 * @param {string} newMessage - New user message
 * @returns {Promise<{response: string, responseTime: number}>}
 */
const chatbotConversation = async (conversationHistory, newMessage) => {
  const startTime = Date.now();

  try {
    const systemPrompt = `You are a helpful medical assistant chatbot for patients.
Provide clear, empathetic, and accurate information.
If asked about symptoms, suggest when to seek emergency care.
Never diagnose - always recommend consulting a healthcare provider for medical advice.
Keep responses under 100 words.`;

    // Build conversation context
    let contextStr = '';
    if (conversationHistory && conversationHistory.length > 0) {
      contextStr = 'Previous conversation:\n';
      conversationHistory.slice(-4).forEach(msg => {
        contextStr += `${msg.role === 'user' ? 'Patient' : 'Assistant'}: ${msg.content}\n`;
      });
      contextStr += '\n';
    }

    const userPrompt = `${contextStr}Patient: ${newMessage}\n\nRespond helpfully and concisely.`;

    const prompt = buildPhi3Prompt(systemPrompt, userPrompt);

    const response = await axios.post(OLLAMA_GENERATE_URL, {
      model: OLLAMA_MODEL,
      prompt: prompt,
      stream: false,
      options: INFERENCE_PARAMS
    }, {
      timeout: 10000
    });

    const responseTime = Date.now() - startTime;

    if (!response.data || !response.data.response) {
      throw new Error('Ollama returned malformed response');
    }

    return {
      response: response.data.response.trim(),
      responseTime
    };

  } catch (error) {
    const responseTime = Date.now() - startTime;
    
    if (error.code === 'ECONNREFUSED' || error.message.includes('unreachable')) {
      return {
        response: 'I apologize, but I am currently unable to respond. The AI service is temporarily unavailable. Please try again in a moment or contact staff for assistance.',
        responseTime,
        error: 'OLLAMA_UNREACHABLE'
      };
    }

    return {
      response: 'I apologize, but I encountered an error processing your message. Please try rephrasing your question or contact staff for assistance.',
      responseTime,
      error: 'INFERENCE_FAILED'
    };
  }
};

module.exports = {
  healthCheck,
  emergencyTriageQuery,
  chatbotConversation,
  buildPhi3Prompt,
  OLLAMA_BASE_URL,
  OLLAMA_MODEL
};
